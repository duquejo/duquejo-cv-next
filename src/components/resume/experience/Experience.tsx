import { cn } from '@/lib/utils';
import type { ExperienceType } from '@/interfaces/experience';
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';

export const Experience = ({
  start_date,
  end_date,
  enterprise,
  isRecent,
  project,
  role,
  resume,
  additional_info,
}: ExperienceType) => {
  return (
    <ol className="first-of-type:pt-0 pt-5 relative border-s border-gray-200">
      <div
        data-test="detail"
        className={cn(
          'absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400',
          isRecent ? 'bg-primary' : 'bg-secondary',
        )}
      />

      {/*Aligned content*/}
      <div className="pb-2 ms-4 flex items-center justify-between lg:justify-normal">
        <span
          data-test="time"
          className={cn(
            'lg:inline-flex border-2 rounded px-2 py-1 font-semibold text-xs text-center',
            isRecent ? 'border-yellow-400' : 'border-secondary',
          )}
        >
          <time className="block">{start_date}</time>
          <time className="block mt-1 lg:mt-0 lg:before:content-['-'] lg:before:mx-1">
            {end_date}
          </time>
        </span>
        <span className="text-xs font-normal ml-2 text-right lg:text-left">{enterprise}</span>
      </div>
      <Card className="ms-4">
        <CardHeader>
          <CardTitle>{project}</CardTitle>
          <CardDescription>{role}</CardDescription>
        </CardHeader>
        {resume && (
          <CardContent>
            {resume.length > 1 ? (
              <ol className="list-disc ml-4 text-justify marker:text-secondary">
                {resume.map((r, i) => (
                  <li key={`resume-${i}`} className="my-1 text-sm text-justify">
                    {r}
                  </li>
                ))}
              </ol>
            ) : (
              <p className="my-1 text-justify text-sm">{resume}</p>
            )}
          </CardContent>
        )}
        {additional_info && (
          <CardFooter>
            <small className="block font-semibold text-justify">{additional_info}</small>
          </CardFooter>
        )}
      </Card>
    </ol>
  );
};
