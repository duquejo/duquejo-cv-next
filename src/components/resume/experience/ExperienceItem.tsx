import { cn } from '@/lib/utils';
import type { ExperienceType } from '@/interfaces/experience.interface';
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';
import { Button } from '@/components/ui/button';
import { Badge } from '@/components/ui/badge';

export const ExperienceItem = ({
  start_date,
  end_date,
  enterprise,
  isRecent,
  project,
  role,
  resume,
  additional_info,
  screenshot,
  url,
}: ExperienceType) => {
  return (
    <ol className="first-of-type:pt-0 pt-5 relative border-s border-gray-200">
      <div
        data-test="detail"
        className={cn(
          'absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400',
          isRecent ? 'bg-primary' : 'bg-secondary',
        )}
      />

      {/*Aligned content*/}
      <div className="pb-2 ms-4 flex items-center justify-between lg:justify-normal">
        <span
          data-test="time"
          className={cn(
            'lg:inline-flex border-2 rounded px-2 py-1 font-semibold text-xs text-center',
            isRecent ? 'border-yellow-400' : 'border-secondary',
          )}
        >
          <time className="block">{start_date}</time>
          <time className="block mt-1 lg:mt-0 lg:before:content-['-'] lg:before:mx-1">
            {end_date}
          </time>
        </span>
        <span className="text-xs font-normal ml-2 text-right lg:text-left">{enterprise}</span>
      </div>
      <Card className="ms-4">
        <CardHeader>
          <CardTitle>{project}</CardTitle>
          <CardDescription>{role}</CardDescription>
        </CardHeader>
        <CardContent className="flex">
          {resume && (
            <>
              <div
                className={cn(
                  'flex flex-col justify-between items-baseline',
                  screenshot && 'basis-6/12',
                )}
              >
                <div className="flex-auto">
                  <span className="font-semibold text-base">Responsibilities</span>
                  {resume.length > 1 ? (
                    <ol className="list-disc ml-4 text-justify marker:text-secondary">
                      {resume.map((r, i) => (
                        <li key={`resume-${i}`} className="my-1 text-sm text-justify">
                          {r}
                        </li>
                      ))}
                    </ol>
                  ) : (
                    <p className="my-1 text-justify text-sm">{resume}</p>
                  )}
                </div>

                {url && (
                  <Button className="mt-5 flex-shrink" variant="secondary">
                    Project&#39;s link
                  </Button>
                )}
              </div>
            </>
          )}
          {screenshot && (
            <picture className="flex flex-grow flex-col px-4 pt-0">
              <div className="aspect-video rounded-xl bg-muted/50" />
            </picture>
          )}
        </CardContent>
        {additional_info && (
          <CardFooter className="flex flex-col items-start gap-y-2">
            <div className="font-semibold text-base">Skills</div>
            <div className="flex gap-1 flex-wrap">
              {additional_info.map((info) => (
                <Badge variant="secondary" key={info}>
                  {info}
                </Badge>
              ))}
            </div>
          </CardFooter>
        )}
      </Card>
    </ol>
  );
};
