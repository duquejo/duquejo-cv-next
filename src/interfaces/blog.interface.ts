import { ReactNode } from 'react';

export interface BlogPost {
  slug: string;
  title: string;
  excerpt: string;
  category: string;
  tags: string[];
  publishDate: string;
  readingTime: string;
  coverImage: string;
}

export interface BlogPostResult {
  Post: () => ReactNode;
  metadata: BlogPost;
}
