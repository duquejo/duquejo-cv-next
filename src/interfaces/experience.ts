export interface ExperienceType {
  start_date: string;
  end_date: string;
  enterprise: string;
  isRecent: boolean;
  project: string;
  role: string;
  resume?: string[];
  additional_info?: string;
}

export type Sizes = 'small' | 'large' | 'medium';
