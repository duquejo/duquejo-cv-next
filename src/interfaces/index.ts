export * from './experience.interface';
export * from './menu-items.interface';
export * from './main-languages.interface';
export * from './services.interface';
export * from './knowledge.interface';
