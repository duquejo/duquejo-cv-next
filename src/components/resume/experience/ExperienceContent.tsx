import { cn } from '@/lib/utils';
import { Button } from '@/components/ui/button';
import { ExperienceType } from '@/interfaces';

export const ExperienceContent = ({ resume, screenshot, url }: Partial<ExperienceType>) => {
  return (
    <div className="flex flex-col lg:flex-row">
      {resume && (
        <div
          className={cn(
            'flex flex-col order-2 lg:order-1 justify-between items-baseline',
            screenshot && 'basis-6/12',
          )}
        >
          <div className="flex-auto">
            <span className="font-semibold text-base">Responsibilities</span>
            {resume.length > 1 ? (
              <ol className="list-disc ml-4 text-justify marker:text-secondary">
                {resume.map((r, i) => (
                  <li key={`resume-${i}`} className="my-1 text-sm text-justify">
                    {r}
                  </li>
                ))}
              </ol>
            ) : (
              <p className="my-1 text-justify text-sm">{resume}</p>
            )}
          </div>
          {url && (
            <Button className="mt-5 lg:flex-shrink lg:w-auto w-full" variant="secondary">
              Project&#39;s link
            </Button>
          )}
        </div>
      )}
      {screenshot && (
        <picture className="order-1 lg:order-2 flex flex-grow flex-col px-0 lg:px-4 pt-0 mb-5 lg:mb-0">
          <div className="aspect-video rounded-xl bg-muted/50" />
        </picture>
      )}
    </div>
  );
};
