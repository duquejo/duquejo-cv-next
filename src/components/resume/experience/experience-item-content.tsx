import { cn } from '@/lib/utils';
import { Button } from '@/components/ui/button';
import { ExperienceType } from '@/interfaces';
import Link from 'next/link';
import Image from 'next/image';

export const ExperienceItemContent = ({
  project = 'Image',
  resume,
  screenshot,
  url,
}: Partial<ExperienceType>) => {
  return (
    <div className="flex flex-col lg:flex-row">
      {resume && (
        <div
          className={cn(
            'flex flex-col order-2 lg:order-1 justify-between items-baseline',
            screenshot && 'basis-6/12',
          )}
        >
          <div className="flex-auto">
            <span className="font-semibold text-base">Responsibilities</span>
            {resume.length > 1 ? (
              <ol className="list-disc ml-4 text-justify marker:text-secondary">
                {resume.map((r, i) => (
                  <li key={`resume-${i}`} className="my-1 text-sm text-justify">
                    {r}
                  </li>
                ))}
              </ol>
            ) : (
              <p className="my-1 text-justify text-sm">{resume}</p>
            )}
          </div>
          {url && (
            <Button className="mt-5 lg:flex-shrink lg:w-auto w-full" variant="secondary" asChild>
              <Link href={url} target="_blank" rel="noopener noreferrer nofollow">
                Project&#39;s link
              </Link>
            </Button>
          )}
        </div>
      )}
      {screenshot && (
        <picture className="order-1 lg:order-2 flex flex-grow flex-col items-center px-0 mb-5 pt-0 lg:px-4 lg:pt-8 lg:mb-0">
          <Image
            src={screenshot}
            width="300"
            height="150"
            loading="lazy"
            alt={project}
            className="rounded object-cover shadow-lg w-auto max-w-xs"
          />
        </picture>
      )}
    </div>
  );
};
