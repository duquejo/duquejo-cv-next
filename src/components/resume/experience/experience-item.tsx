import { cn } from '@/lib/utils';
import type { ExperienceType } from '@/interfaces/experience.interface';
import {
  Card,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import { ExperienceItemContent } from '@/components/resume/experience/experience-item-content';

export const ExperienceItem = ({
  start_date,
  end_date,
  enterprise,
  isRecent,
  project,
  role,
  resume,
  additional_info,
  screenshot,
  url,
}: ExperienceType) => {
  return (
    <ol className="first-of-type:pt-0 pt-5 relative border-s border-gray-200">
      <div
        data-testid="detail"
        className={cn(
          'absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-gray-400',
          isRecent ? 'bg-primary' : 'bg-secondary',
        )}
      />

      {/*Aligned content*/}
      <div className="pb-2 ms-4 flex items-center justify-between lg:justify-normal">
        <span
          data-testid="time"
          className={cn(
            'lg:inline-flex items-center border-2 rounded px-2 py-1 font-semibold text-xs text-center',
            isRecent ? 'border-primary' : 'border-secondary',
          )}
        >
          <time className="block">{start_date}</time>
          {end_date && (
            <time className="block lg:before:content-['-'] lg:before:mx-1">{end_date}</time>
          )}
        </span>
        {enterprise && (
          <span className="text-xs font-normal ml-2 text-right lg:text-left">{enterprise}</span>
        )}
      </div>
      <Card className="ms-4">
        <CardHeader>
          <CardTitle>{project}</CardTitle>
          <CardDescription>{role}</CardDescription>
        </CardHeader>
        <CardContent className="py-0">
          <ExperienceItemContent
            project={project}
            resume={resume}
            screenshot={screenshot}
            url={url}
          />
        </CardContent>
        {additional_info && (
          <CardFooter className="flex flex-col items-start gap-y-2 pt-5">
            <div className="font-semibold text-base">Skills</div>
            <div className="flex lg:gap-1 gap-2 flex-wrap lg:justify-normal justify-around">
              {additional_info.map((info) => (
                <Badge variant="secondary" key={info}>
                  {info}
                </Badge>
              ))}
            </div>
          </CardFooter>
        )}
      </Card>
    </ol>
  );
};
