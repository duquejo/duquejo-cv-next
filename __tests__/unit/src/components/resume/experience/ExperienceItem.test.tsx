import { ExperienceType } from '@/interfaces';
import { render, screen } from '@testing-library/react';
import { ExperienceItem } from '@/components/resume/experience/ExperienceItem';

describe('<ExperienceItemContent /> tests', () => {
  const data: ExperienceType = {
    resume: ['foo', 'bar', 'baz'],
    screenshot: '/foo/bar.webp',
    url: 'https://foo.bar',
    enterprise: 'John Doe',
    additional_info: ['Next JS', 'React', 'Vitest'],
    isRecent: false,
    start_date: '2025-01-01',
    end_date: '2025-01-02',
    role: 'Software analyst',
    project: 'Foo-Bar Baz Inc.',
  };

  it('Should match the snapshot with the default args', () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { resume, additional_info, screenshot, url, end_date, ...mandatoryFields } = data;

    const { container } = render(<ExperienceItem {...mandatoryFields} />);

    expect(container).toMatchSnapshot();

    expect(screen.getByTestId('time')).not.toHaveTextContent(data.end_date!);
    expect(screen.getByTestId('time')).not.toHaveClass('border-yellow-400');

    expect(screen.getByTestId('detail')).not.toHaveClass('bg-primary');
  });

  it('Should match the snapshot with the available args', () => {
    const { container } = render(<ExperienceItem {...data} />);

    expect(container).toMatchSnapshot();

    expect(screen.getByText('Skills')).toBeInTheDocument();
    data.additional_info!.forEach((item) => expect(screen.getByText(item)).toBeInTheDocument());

    expect(screen.getByTestId('time')).toHaveTextContent(data.start_date);
    expect(screen.getByTestId('time')).toHaveTextContent(data.end_date!);
  });

  it('Should stylize the time tags if the parameter "isRecent" is truthy', () => {
    const isRecentWithData = {
      ...data,
      isRecent: true,
    };

    render(<ExperienceItem {...isRecentWithData} />);

    expect(screen.getByTestId('time')).toHaveClass('border-yellow-400');
    expect(screen.getByTestId('detail')).toHaveClass('bg-primary');
  });
});
