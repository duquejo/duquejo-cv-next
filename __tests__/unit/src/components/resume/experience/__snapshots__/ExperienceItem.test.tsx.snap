// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ExperienceItemContent /> tests > Should match the snapshot with the available args 1`] = `
<div>
  <ol
    class="first-of-type:pt-0 pt-5 relative border-s border-gray-200"
  >
    <div
      class="absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-secondary"
      data-testid="detail"
    />
    <div
      class="pb-2 ms-4 flex items-center justify-between lg:justify-normal"
    >
      <span
        class="lg:inline-flex items-center border-2 rounded px-2 py-1 font-semibold text-xs text-center border-secondary"
        data-testid="time"
      >
        <time
          class="block"
        >
          2025-01-01
        </time>
        <time
          class="block lg:before:content-['-'] lg:before:mx-1"
        >
          2025-01-02
        </time>
      </span>
      <span
        class="text-xs font-normal ml-2 text-right lg:text-left"
      >
        John Doe
      </span>
    </div>
    <div
      class="rounded-xl border bg-card text-card-foreground shadow ms-4"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <div
          class="font-semibold leading-none tracking-tight"
        >
          Foo-Bar Baz Inc.
        </div>
        <div
          class="text-sm text-muted-foreground"
        >
          Software analyst
        </div>
      </div>
      <div
        class="p-6 py-0"
      >
        <div
          class="flex flex-col lg:flex-row"
        >
          <div
            class="flex flex-col order-2 lg:order-1 justify-between items-baseline basis-6/12"
          >
            <div
              class="flex-auto"
            >
              <span
                class="font-semibold text-base"
              >
                Responsibilities
              </span>
              <ol
                class="list-disc ml-4 text-justify marker:text-secondary"
              >
                <li
                  class="my-1 text-sm text-justify"
                >
                  foo
                </li>
                <li
                  class="my-1 text-sm text-justify"
                >
                  bar
                </li>
                <li
                  class="my-1 text-sm text-justify"
                >
                  baz
                </li>
              </ol>
            </div>
            <a
              class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80 h-9 px-4 py-2 mt-5 lg:flex-shrink lg:w-auto w-full"
              href="https://foo.bar"
              rel="noopener noreferrer nofollow"
              target="_blank"
            >
              Project's link
            </a>
          </div>
          <picture
            class="order-1 lg:order-2 flex flex-grow flex-col items-center px-0 mb-5 pt-0 lg:px-4 lg:pt-8 lg:mb-0"
          >
            <img
              alt="Foo-Bar Baz Inc."
              class="rounded object-cover shadow-lg w-auto max-w-xs"
              data-nimg="1"
              decoding="async"
              height="150"
              loading="lazy"
              src="/_next/image?url=%2Ffoo%2Fbar.webp&w=640&q=75"
              srcset="/_next/image?url=%2Ffoo%2Fbar.webp&w=384&q=75 1x, /_next/image?url=%2Ffoo%2Fbar.webp&w=640&q=75 2x"
              style="color: transparent;"
              width="300"
            />
          </picture>
        </div>
      </div>
      <div
        class="p-6 flex flex-col items-start gap-y-2 pt-5"
      >
        <div
          class="font-semibold text-base"
        >
          Skills
        </div>
        <div
          class="flex lg:gap-1 gap-2 flex-wrap lg:justify-normal justify-around"
        >
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80"
          >
            Next JS
          </div>
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80"
          >
            React
          </div>
          <div
            class="inline-flex items-center rounded-md border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent bg-secondary text-secondary-foreground hover:bg-secondary/80"
          >
            Vitest
          </div>
        </div>
      </div>
    </div>
  </ol>
</div>
`;

exports[`<ExperienceItemContent /> tests > Should match the snapshot with the default args 1`] = `
<div>
  <ol
    class="first-of-type:pt-0 pt-5 relative border-s border-gray-200"
  >
    <div
      class="absolute w-2 h-2 rounded-full mt-1.5 -start-1 bg-secondary"
      data-testid="detail"
    />
    <div
      class="pb-2 ms-4 flex items-center justify-between lg:justify-normal"
    >
      <span
        class="lg:inline-flex items-center border-2 rounded px-2 py-1 font-semibold text-xs text-center border-secondary"
        data-testid="time"
      >
        <time
          class="block"
        >
          2025-01-01
        </time>
      </span>
      <span
        class="text-xs font-normal ml-2 text-right lg:text-left"
      >
        John Doe
      </span>
    </div>
    <div
      class="rounded-xl border bg-card text-card-foreground shadow ms-4"
    >
      <div
        class="flex flex-col space-y-1.5 p-6"
      >
        <div
          class="font-semibold leading-none tracking-tight"
        >
          Foo-Bar Baz Inc.
        </div>
        <div
          class="text-sm text-muted-foreground"
        >
          Software analyst
        </div>
      </div>
      <div
        class="p-6 py-0"
      >
        <div
          class="flex flex-col lg:flex-row"
        />
      </div>
    </div>
  </ol>
</div>
`;
